sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -17,61 +17,179 @@
+                  aggNode // return original plan node
+                } else {
+                  // use the aggregate columns as the output columns
+                  // e.g. TABLE t (c1 INT, c2 INT, c3 INT)
+                  // SELECT min(c1), max(c1) FROM t;
+                  // Use min(c1), max(c1) as output for DataSourceV2ScanRelation
+                  // We want to have the following logical plan:
+                  // == Optimized Logical Plan ==
+                  // Aggregate [min(min(c1)#21) AS min(c1)#17, max(max(c1)#22) AS max(c1)#18]
+                  // +- RelationV2[min(c1)#21, max(c1)#22] parquet file ...
+                  val output = aggregates.map {
+                    case agg: AggregateExpression =>
+                      AttributeReference(toPrettySQL(agg), agg.dataType)()
+                  }
+
+                  // No need to do column pruning because only the aggregate columns are used as
+                  // DataSourceV2ScanRelation output columns. All the other columns are not
+                  // included in the output. Since PushDownUtils.pruneColumns is not called,
+                  // ScanBuilder.requiredSchema is not pruned, but ScanBuilder.requiredSchema is
+                  // not used anyways. The schema for aggregate columns will be built in Scan.
+                  val scan = sHolder.builder.build()
+
+                  logInfo(
+                    s"""
+                       |Pushing operators to ${sHolder.relation.name}
+                       |Pushed Aggregate Functions:
+                       | ${pushedAggregates.aggregateExpressions.mkString(", ")}
+                       |Output: ${output.mkString(", ")}
+                      """.stripMargin)
+
+                  val scanRelation = DataSourceV2ScanRelation(sHolder.relation, scan, output)
+                  val plan = Aggregate(groupingExpressions, resultExpressions, scanRelation)
+
+                  // Change the optimized logical plan to reflect the pushed down aggregate
+                  // e.g. TABLE t (c1 INT, c2 INT, c3 INT)
+                  // SELECT min(c1), max(c1) FROM t;
+                  // The original logical plan is
+                  // Aggregate [min(c1#9) AS min(c1)#17, max(c1#9) AS max(c1)#18]
+                  // +- RelationV2[c1#9] parquet ...
+                  //
+                  // After change the V2ScanRelation output to [min(_1)#21, max(_1)#22]
+                  // we have the following
+                  // !Aggregate [min(_1#9) AS min(_1)#17, max(_1#9) AS max(_1)#18]
+                  // +- RelationV2[min(_1)#21, max(_1)#22] parquet ...
+                  //
+                  // We want to change it to
+                  // == Optimized Logical Plan ==
+                  // Aggregate [min(min(c1)#21) AS min(c1)#17, max(max(c1)#22) AS max(c1)#18]
+                  // +- RelationV2[min(c1)#21, max(c1)#22] parquet file ...
+                  var i = 0
+                  plan.transformExpressions {
```

Comment:
one thing we should take care of is the changing of expr ID. For example,
```
Sort(a#1)
  Aggregate(groupBy=a#1, result=a#1, sum(b#2))
    V2Relation(a#1, b#2)
```
After pushdown
```
!Sort(a#1)
  Aggregate(groupBy=a#3, result=a#3, sum(sum(b)#4))
    V2Relation(a#3, sum(b)#4)
```

The sort operator becomes invalid because `a#1` is missing now. We should find a way to keep the original expr ID, via `Alias(...)(exprId = originalId)`

Reply:
nvm, the output is kept: https://github.com/apache/spark/pull/32049/files#diff-cf61a7ed875ef854164cd3d741e5adc4c1694c05b1bb3bdc19dd5c8dc5bca8bdR117